### ISSUE TITLE
Transforms

### ISSUE THREAD

**Issue by samdj** [knex/knex#1443]
I'd like to be able to transform the returned data without actually converting the query into a promise.

``` js
function getStuff() { return Promise.resolve('stuff') )

const query = knex
  .table('test')
  .select('*')
  .transform((rows) => {
    return Promise.map(rows, async (row) => {
      row.stuff = await getStuff(row.id)
      return row
    )
  })

// ... later

query.where({ id: 22 }).limit(33).then((rows) => {
  rows[0].stuff // 'stuff'
})
```

The important part of this is that one can return the query to be further modified, while maintaining the transformed output and queuing up transformed promises for when the query is properly executed.

Internally it could just mean:
- Storing `.transform` callbacks in an array
- When `.then` and other executing calls are made:
  - Run `Promise.reduce` over the transforms and `.then` that

Happy to make a pull request if it's likely to be added.

**Comment by tgriesser** [comment]
This will be added in 1.0 with a concept of `hooks`, see #1659

```js
const query = knex
  .table('test')
  .select('*')
  .hook('transformResponse', (rows) => {
    return Promise.map(rows, async (row) => {
      row.stuff = await getStuff(row.id)
      return row
    )
  })
```

Still need to figure out the semantics around what happens if you try to attach more than one hook... do they pipe results into each other? Does a later hook replace an earlier hook?

**Comment by nfour** [comment]
That's sweet @tgriesser !

I would intuit that they would queue and chain together synchronously by default.

Maybe also add an alternate syntax  `.parallelHook`, or a 3rd argument to force parallel resolution.

**Comment by wubzz** [comment]
I think `postProcessResponse` covers this, except it might not support async functions..

**Comment by atmortensen** [comment]
Is there any way to add a `postProcessResponse` to an individual query instead of the whole connection?

**Comment by elhigu** [comment]
no, I don't think so... except maybe with help of http://knexjs.org/#Builder-queryContext to use some variable to pass custom postProcessResponse function to the query.

**Comment by ftaiolivista** [comment]
> I think `postProcessResponse` covers this, except it might not support async functions..

Any way to have something like postProcessResponse but with async support?

**Comment by mercmobily** [comment]
Thank you for raising this feature request about adding a `.then()` middleware for transforming query results. This is an excellent suggestion that would provide much more flexibility in processing responses on a per-query basis.

The good news is that Knex has addressed the core requirement of this issue through the `postProcessResponse` configuration option, which has been available since around version 0.19.0. This feature allows you to transform query responses, and I can confirm through testing that it **does support async functions**, contrary to some speculation in the thread.

Here's a working example demonstrating async postProcessResponse functionality:

```javascript
const knex = require('knex')({
  client: 'sqlite3',
  connection: ':memory:',
  postProcessResponse: async (result, queryContext) => {
    // Async operations work perfectly fine
    await someAsyncOperation();
    
    if (Array.isArray(result)) {
      return Promise.all(result.map(async (row) => {
        row.extraData = await fetchAdditionalData(row.id);
        return row;
      }));
    }
    return result;
  }
});
```

I've verified this works correctly in the current codebase at lib/client.js:170-175 where postProcessResponse is called, and the runner properly awaits the response at lib/execution/runner.js:153-156.

**Limitations of Current Implementation**

Even though postProcessResponse addresses the basic need, it has significant limitations compared to your proposed solution:

1. **Global scope only**: The postProcessResponse is configured at the connection level, not per-query. This means all queries through that connection get the same processing, which isn't ideal when different queries need different transformations.

2. **No chaining**: Unlike your proposed `.hook()` method, you cannot chain multiple transformations or conditionally apply them to specific queries.

3. **Limited context**: Whilst you can use `queryContext()` to pass custom data to help conditionally process responses, this is a workaround rather than a proper per-query solution.

**The Hooks Proposal**

The maintainer @tgriesser mentioned in 2016 that a more comprehensive hooks system was planned for version 1.0 (referenced in issue #1659), which would have provided exactly what you're asking for:

```javascript
knex.table('test')
  .select('*')
  .hook('transformResponse', async (rows) => {
    return Promise.map(rows, async (row) => {
      row.stuff = await getStuff(row.id);
      return row;
    });
  });
```

Unfortunately, this hooks system hasn't been implemented yet, even though Knex has progressed well beyond version 1.0. The concept remains valuable and would provide much better flexibility than the current global postProcessResponse approach.


Until proper hooks are implemented, you can achieve per-query processing by using queryContext creatively:

```javascript
// Define your transformations
const transformations = {
  addUserDetails: async (rows) => { /* ... */ },
  formatDates: (rows) => { /* ... */ }
};

// Configure postProcessResponse to use queryContext
const knex = require('knex')({
  postProcessResponse: async (result, queryContext) => {
    if (queryContext?.transform && transformations[queryContext.transform]) {
      return transformations[queryContext.transform](result);
    }
    return result;
  }
});

// Use per-query transformation
const users = await knex('users')
  .queryContext({ transform: 'addUserDetails' })
  .select('*');
```

This pattern gives you some of the flexibility you're looking for, even though it's not as elegant as the proposed hooks system would be.

**Recent Developments**

There has been significant recent activity around implementing a comprehensive hooks system for Knex:

1. **PR #6070 (May 2024)**: An open pull request adds a `preExecHook` capability for pre-SQL execution actions, complementing the existing postProcessResponse functionality.

2. **Issue #6270 (August 2025)**: A comprehensive proposal for a Unified Hook System that would provide exactly the per-query transformation capability you requested. This proposal includes hooks at multiple lifecycle points including `beforeCompile`, `afterCompile`, `beforeExecute`, and `afterExecute`, with both global and per-query registration options.

3. **PR #5541 (Merged)**: Migration lifecycle hooks have already been successfully implemented, showing that the maintainers are actively working on expanding hook capabilities in Knex.

The proposed unified hook system (#6270) would allow exactly what you originally requested:

```javascript
// Per-query transformation hooks
knex('users')
  .afterExecute(async (response, context) => {
    // Transform response for this specific query
    return Promise.map(response, async (row) => {
      row.additionalData = await fetchData(row.id);
      return row;
    });
  })
  .select('*');
```

**Conclusion**

The core functionality you requested exists through postProcessResponse with full async support, but the per-query hooks system would be a valuable addition to Knex. The workaround using queryContext provides a path forward for those needing per-query transformations today. Given the active development on PR #6070 and the comprehensive proposal in #6270, it appears that a proper per-query hooks system is likely to be implemented in the near future, finally delivering on the promise made back in 2016.

I am closing this in favour of #6070

**Related Issues**
- #1659 - Original hooks proposal for v1.0
- #3931 - postProcessResponse support for streams
- #3040 - SQLite3 compatibility with postProcessResponse
- #3111 - queryContext passing to raw queries
- #6070 - PR adding preExecHook capability
- #6270 - Unified Hook System proposal for query lifecycle
- #5541 - Migration lifecycle hooks (merged)
